temporal-sdk/src/main/java/io/temporal/internal/replay/ReplayWorkflowTaskHandler.java

```diff
@@ -101,20 +101,119 @@ public ReplayWorkflowTaskHandler(
     String workflowType = workflowTask.getWorkflowType().getName();
     Scope metricsScope =
         options.getMetricsScope().tagged(ImmutableMap.of(MetricsTag.WORKFLOW_TYPE, workflowType));
-    try {
-      if (workflowTask.hasQuery()) {
-        // Legacy query codepath
+    if (workflowTask.hasQuery()) {
+      // Legacy query codepath
+      try {
         return handleQueryOnlyWorkflowTask(workflowTask.toBuilder(), metricsScope);
-      } else {
+      } catch (Throwable e) {
+        return createQueryResult(workflowTask, Optional.empty(), e);
+      }
+    } else {
+      try {
         return handleWorkflowTaskWithEmbeddedQuery(workflowTask.toBuilder(), metricsScope);
+      } catch (Throwable e) {
+        metricsScope.counter(MetricsType.WORKFLOW_TASK_EXECUTION_FAILURE_COUNTER).inc(1);
+        return failureToWFTResult(workflowTask, e);
       }
+    }
+  }
+
+  private Result handleWorkflowTaskWithEmbeddedQuery(
+      PollWorkflowTaskQueueResponse.Builder workflowTask, Scope metricsScope) throws Throwable {
+    AtomicBoolean createdNew = new AtomicBoolean();
+    WorkflowExecution execution = workflowTask.getWorkflowExecution();
+    WorkflowRunTaskHandler workflowRunTaskHandler = null;
+    boolean useCache = stickyTaskQueueName != null;
+
+    try {
+      workflowRunTaskHandler =
+          getOrCreateWorkflowExecutor(useCache, workflowTask, metricsScope, createdNew);
+      WorkflowTaskResult result = workflowRunTaskHandler.handleWorkflowTask(workflowTask);
+      if (result.isFinalCommand()) {
+        cache.invalidate(execution, metricsScope, "FinalCommand", null);
+      } else if (useCache && createdNew.get()) {
+        cache.addToCache(execution, workflowRunTaskHandler);
+      }
+      return createCompletedWFTRequest(
+          workflowTask.getWorkflowType().getName(), workflowTask, result);
     } catch (Throwable e) {
-      metricsScope.counter(MetricsType.WORKFLOW_TASK_EXECUTION_FAILURE_COUNTER).inc(1);
-      return failureToResult(workflowTask, e);
+      // Note here that the executor might not be in the cache, even when the caching is on. In that
+      // case we need to close the executor explicitly. For items in the cache, invalidation
+      // callback will try to close again, which should be ok.
+      if (workflowRunTaskHandler != null) {
+        workflowRunTaskHandler.close();
+      }
+
+      if (useCache) {
+        cache.invalidate(execution, metricsScope, "Exception", e);
+        // If history if full and exception occurred then sticky session hasn't been established
+        // yet and we can avoid doing a reset.
+        if (!isFullHistory(workflowTask)) {
+          resetStickyTaskQueue(execution);
+        }
+      }
+      throw e;
+    } finally {
+      if (useCache) {
+        cache.markProcessingDone(execution);
+      } else if (workflowRunTaskHandler != null) {
+        workflowRunTaskHandler.close();
```

Comment:
Is this safe for calling multiple times? You may call this in the catch block and again here.

Reply:
Yeah, it is. I will port a comment here.